### PR TITLE
Added per shape attribution

### PR DESCRIPTION
**Pull request by kdiluca**
# Issue

Adding new per shape attribution to the attributes controller.  This PR will also calculate the distance & duration between each shape point by walk along the shape.

## Tasklist

 - [x] Add tests
 - [ ] Review - you must request approval to merge any PR to master
 - [ ] Add #fixes with the issue number that this PR addresses
 - [ ] Generally use squash merge to rebase and clean comments before merging
 - [ ] Update the [changelog](CHANGELOG.md)
 - [ ] Update relevant [documentation](docs/README.md)

## Requirements / Relations

 Link any requirements here. Other pull requests this PR is based on?

### CODE REVIEW

File: src/thor/triplegbuilder.cc [hunk 1]
```diff
@@ -766,6 +808,10 @@ TripLegBuilder::Build(const AttributesController& controller,
   const DirectedEdge* prev_de = nullptr;
   // TODO: this is temp until we use transit stop type from transitland
   TransitPlatformInfo_Type prev_transit_node_type = TransitPlatformInfo_Type_kStop;
+
+  if (controller.category_attribute_enabled(kShapeAttributesCategory)) {
```
Comment (kevinkreiser) [inline src/thor/triplegbuilder.cc]: instead of having this `if` twice you could move it to the beginning of the function before the special case single edge route

File: test/shape_attributes.cc [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+} // namespace
+
+int main() {
+  test::suite suite("shape_attributes");
+
+  suite.test(TEST_CASE(test_attributes));
+
+  suite.test(TEST_CASE(test_interrupt));
```
Comment (kevinkreiser) [inline test/shape_attributes.cc]: you dont need the interrupt test at all, please remove

File: test/shape_attributes.cc [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+      "action":"include"}})");
+  actor.cleanup();
+  auto attributes_json = actor.trace_attributes(R"({"shape":[{"lat":40.546115,"lon":-76.385076},
+      {"lat":40.544232,"lon":-76.385752}],"costing":"auto","shape_match":"walk_or_snap",
+      "filters":{"attributes":["shape_attributes.time","shape_attributes.length"],
+      "action":"include"}})");
+  actor.cleanup();
+  auto attributes = json_to_pt(attributes_json);
+  attributes_json.find("length");
+  attributes_json.find("time");
```
Comment (kevinkreiser) [inline test/shape_attributes.cc]: this test doesnt make sense to me. it calls trace attributes twice, the first time complete ignoring the result, could see maybe doing that in the general actor test to make sure that results dont bleed through or something but it doesnt make sense here.

additoinally this test doesnt test anything. it parses the response but then doesnt validate anything. you should make sure you have a multi edge route, you should also decode the shape and measure that the length between each is appx the same as the length in the shape attributes. you can then also ask for the edge time and validate that that makes sense as well.

File: proto/trip.proto [hunk 4]
```diff
@@ -240,10 +240,10 @@ message TripLeg {
       kMotorwayJunction = 9;        // Highway = motorway_junction
       kBorderControl = 10;          // Border control
     }
-    
+
     optional Edge edge = 1;
-    repeated IntersectingEdge intersecting_edge = 2; 
-    optional uint32 elapsed_time = 3;        // seconds from start of leg
+    repeated IntersectingEdge intersecting_edge = 2;
+    optional double elapsed_time = 3;        // milliseconds from start of leg
```
Comment (kevinkreiser) [inline proto/trip.proto]: units didnt change

File: src/thor/triplegbuilder.cc [hunk 5]
```diff
@@ -161,6 +165,46 @@ void AssignAdmins(const AttributesController& controller,
...
+                        const DirectedEdge* directededge,
+                        const std::shared_ptr<sif::DynamicCost>& costing,
+                        std::vector<PointLL>::const_iterator shape_begin,
+                        std::vector<PointLL>::const_iterator shape_end,
+                        TripLeg& trip_path) {
+  if (trip_path.has_shape_attributes()) {
+    // calculates total edge time and total edge length
+    double edge_time =
+        costing->EdgeCost(directededge, graphtile->GetSpeed(directededge)).secs; // seconds
+    double edge_length = directededge->length();                                 // meters
```
Comment (kevinkreiser) [inline src/thor/triplegbuilder.cc]: TODO: get the measured length from shape (full shape) to increase precision

File: src/tyr/trace_serializer.cc [hunk 6]
```diff
@@ -19,6 +19,9 @@ constexpr size_t kConfidenceScoreIndex = 0;
 constexpr size_t kRawScoreIndex = 1;
 constexpr size_t kMatchResultsIndex = 2;
 constexpr size_t kTripLegIndex = 3;
+constexpr double MILLISECOND_TO_SEC = 0.001;
```
Comment (kevinkreiser) [inline src/tyr/trace_serializer.cc]: we have similar constants in midgard/constants.h move missing ones there. feel free to also make double versions of these if it suits your fancy

File: test/shape_attributes.cc [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+      sum_lengths += shape_attributes_length[j].GetDouble();
+
+      if (!midgard::equal(edge_speed, shape_attributes_speed[j].GetDouble(), .15)) {
+        throw std::logic_error("Expected shape_attributes.speed to approx equal " +
+                               std::to_string(shape_attributes_speed[j].GetFloat()) +
+                               " | Found: " + std::to_string(edge_speed));
+      }
+    }
+
+    // Can't assert that sum of shape times equals edge's elapsed_time because elapsed_time includes
```
Comment (kevinkreiser) [inline test/shape_attributes.cc]: it would be cool to add a test case where we can be sure turn costs wont impact. pick a long edge and put a point at the beginning and end. it wont go t hrough any intersections so you wont incur any turn costs. this also lets us test the single edge special case.
